File: solutions/python/secret-handshake/3/secret_handshake.py
"""Secret handshake command decoder."""

ACTIONS = {
    1: "wink",
    2: "double blink",
    4: "close your eyes",
    8: "jump",
}

REVERSE = 16


def commands(binary_str):
    """Return the handshake commands encoded by a binary string."""
    value = int(binary_str, 2)
    result = [action for bit, action in ACTIONS.items() if has_bit(value, bit)]

    return result[::-1] if has_bit(value, REVERSE) else result


def has_bit(value, bit):
    """Return whether the bit mask is set in value."""
    return value & bit

File: solutions/fsharp/secret-handshake/2/SecretHandshake.fs
module SecretHandshake

let private actions =
    [
        1, "wink"
        2, "double blink"
        4, "close your eyes"
        8, "jump"
    ]

let private reverseAction = 16

let private isSet number mask = number &&& mask <> 0

let private actionIfSet number (mask, action) =
    if isSet number mask then Some action else None

let commands number =
    let selected = actions |> List.choose (actionIfSet number)

    if isSet number reverseAction then List.rev selected else selected
